### PR TITLE
Added http(s)/socks proxy support

### LINKED ISSUE #980: HTTP error code 403 on instaloader.Post.from_shortcode()
**Describe the bug**
unable to make post object using instaloader.Post.from_shortcode() in python 3.8

**To Reproduce**
```python
from instaloader import Instaloader, Post

L = Instaloader()
L.load_session_from_file('username', 'session_name')
post = Post.from_shortcode(L.context, 'SHORTCODE')
print(post.url)

```

**Expected behavior**
I wanted to make a true Post object from an instagram post shortcode

**Error messages and tracebacks**
```
Loaded session from session_name.
JSON Query to graphql/query: HTTP error code 403. [retrying; skip with ^C]
JSON Query to graphql/query: HTTP error code 403. [retrying; skip with ^C]
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 339, in get_json
    raise ConnectionException("HTTP error code {}.".format(resp.status_code))
instaloader.exceptions.ConnectionException: HTTP error code 403.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 339, in get_json
    raise ConnectionException("HTTP error code {}.".format(resp.status_code))
instaloader.exceptions.ConnectionException: HTTP error code 403.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 339, in get_json
    raise ConnectionException("HTTP error code {}.".format(resp.status_code))
instaloader.exceptions.ConnectionException: HTTP error code 403.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "insta.py", line 5, in <module>
    post = Post.from_shortcode(L.context, 'CF0KFB8pWsa')
  File "/usr/local/lib/python3.8/dist-packages/instaloader/structures.py", line 83, in from_shortcode
    post._node = post._full_metadata
  File "/usr/local/lib/python3.8/dist-packages/instaloader/structures.py", line 157, in _full_metadata
    self._obtain_metadata()
  File "/usr/local/lib/python3.8/dist-packages/instaloader/structures.py", line 144, in _obtain_metadata
    pic_json = self._context.graphql_query(
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 416, in graphql_query
    resp_json = self.get_json('graphql/query',
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 382, in get_json
    return self.get_json(path=path, params=params, host=host, session=sess, _attempt=_attempt + 1)
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 382, in get_json
    return self.get_json(path=path, params=params, host=host, session=sess, _attempt=_attempt + 1)
  File "/usr/local/lib/python3.8/dist-packages/instaloader/instaloadercontext.py", line 372, in get_json
    raise ConnectionException(error_string) from err
instaloader.exceptions.ConnectionException: JSON Query to graphql/query: HTTP error code 403.
```

**Instaloader version**
4.6

**Additional context**
Thanks for your great works and support! <3

### PR DESCRIPTION
**Pull request by realSamy**
- Fixes #980 - In case client IP is blocked/limited by Instagram
 - More general: Added HTTP(S)/Socks5 proxy support to be used in connections to Instagram
 - As I tested it is ready to merge and get used

### CODE REVIEW

**Review by Thammus — COMMENTED** [review]
Thanks a lot for this PR. Proxy support is a feature that has been suggested multiple times by many users.
The commented-out function `set_proxies()` contains a great description of how to use the proxy feature. It would be nice to have a note on the usage of the parameter in the Instaloader Class.

File: instaloader/instaloader.py [hunk 1]
```diff
@@ -187,10 +187,14 @@ def __init__(self,
                  rate_controller: Optional[Callable[[InstaloaderContext], RateController]] = None,
                  resume_prefix: Optional[str] = "iterator",
                  check_resume_bbd: bool = True,
-                 slide: Optional[str] = None):
+                 slide: Optional[str] = None,
+                 proxies=None):
```
[the remark this answers is not in the document]
Comment (realSamy) [reply instaloader/instaloader.py:191]: I tried so many typings (e.g. Mutable Mapping, Dict, etc) but each time got some confusions in mypy or pylint checks about usage of the parameter in other cases, so I used it in this way :(

File: instaloader/instaloader.py [hunk 2]
```diff
@@ -187,10 +187,14 @@ def __init__(self,
...
 
+        if proxies is None:
+            proxies = {'http': None, 'https': None}
```
[the remark this answers is not in the document]
Comment (realSamy) [reply instaloader/instaloader.py:194]: Uh I didn't know that, will fix it in another commit

File: instaloader/instaloader.py [hunk 3]
```diff
@@ -465,6 +470,19 @@ def download_hashtag_profilepic(self, hashtag: Hashtag) -> None:
...
+    #     :param proxies: A :class:`MutableMapping` dict of proxies in following format:
+    #      {'http': 'socks://proxy.host:port', 'https': 'https://proxy.host:port'}
+    #      you can use following protocols for each field: http, https, socks5.
+    #     """
+    #     self.proxies = proxies
+    #
+    # def get_proxies(self) -> dict:
+    #     """Get current using HTTP(S)/Socks5 proxy."""
+    #     return self.proxies
+
```
[the remark this answers is not in the document]
Comment (realSamy) [reply instaloader/instaloader.py:485]: I got some mypy or pylint errors about `set_proxies`' descriptions, I couldn't find out why it doesn't work, and another thing that should be fixed is **instaloader.Instaloader.context** should be recreated after calling `set_proxies`, will fix it in next commit then

**Review by Thammus — COMMENTED** [review]
The target branch for new features should be `upcoming/v4.7` instead of `master`.

Since requests already supports proxies through environment variables, could you explain what is the advantage of having a `proxies` attribute in InstaloaderContext?

**Comment by bbepis** [comment]
Could this have an argument added to `__main__.py`? Right now it can only be used by hardcoding it the proxy values, or using an alternative script to invoke the code. Otherwise I have been using these changes myself and they work very well.

**Comment by Thammus** [comment]
@bbepis Feel free to create a Pull Request that introduces this functionality for the CLI.

**Comment by bbepis** [comment]
@Thammus is it possible for me to make a PR while this remains unmerged? or do you mean make a PR into your personal repo?

**Comment by Thammus** [comment]
> @Thammus is it possible for me to make a PR while this remains unmerged?

Yes, for example you can base your changes on [`realSamy:master`](https://github.com/realSamy/instaloader). The interface, i.e. the parameter `proxy` of the Instaloader class, will most likely not change.

**Comment by aandergr** [comment]
I accidentally closed this pull request by removing the branch it based on (originally `upcoming/v4.7`). It is now reopened and based on the new `upcoming/v4.8` branch. Sorry for the inconvenience.

**Pull request closed without merging**